File: ios/CoverColor.swift
//
//  CoverColor.swift
//  react-native-cover
//
//  UIColor hex parsing for the public Cover API.
//
//  Accepts `#RGB`, `#RGBA`, `#RRGGBB`, and `#RRGGBBAA`. The 4- and 8-char
//  forms are RGB-then-alpha (CSS convention), matching the existing
//  public API. Names like "red" and the AARRGGBB ordering are
//  intentionally rejected so iOS and Android behave identically.
//

import UIKit

extension UIColor {
  convenience init?(coverHex: String) {
    var s = coverHex.trimmingCharacters(in: .whitespacesAndNewlines)
    if s.hasPrefix("#") { s.removeFirst() }

    let expanded: String
    switch s.count {
    case 3:
      // #RGB → #RRGGBB
      expanded = s.map { "\($0)\($0)" }.joined()
    case 4:
      // #RGBA → #RRGGBBAA
      expanded = s.map { "\($0)\($0)" }.joined()
    case 6, 8:
      expanded = s
    default:
      return nil
    }

    var rgba: UInt64 = 0
    guard Scanner(string: expanded).scanHexInt64(&rgba) else { return nil }

    let r, g, b, a: CGFloat
    if expanded.count == 6 {
      r = CGFloat((rgba & 0xFF0000) >> 16) / 255
      g = CGFloat((rgba & 0x00FF00) >> 8) / 255
      b = CGFloat(rgba & 0x0000FF) / 255
      a = 1.0
    } else {
      r = CGFloat((rgba & 0xFF000000) >> 24) / 255
      g = CGFloat((rgba & 0x00FF0000) >> 16) / 255
      b = CGFloat((rgba & 0x0000FF00) >> 8) / 255
      a = CGFloat(rgba & 0x000000FF) / 255
    }
    self.init(red: r, green: g, blue: b, alpha: a)
  }
}

File: android/src/main/java/com/margelo/nitro/cover/CoverColor.kt
package com.margelo.nitro.cover

import android.graphics.Color

/**
 * Accepts `#RGB`, `#RGBA`, `#RRGGBB`, and `#RRGGBBAA`. The 4- and 8-char
 * forms are RGB-then-alpha (CSS convention), matching the iOS
 * implementation. Names like `"red"` and the AARRGGBB ordering are
 * intentionally rejected so iOS and Android behave identically.
 *
 * Returns null for unparsable input — callers fall back to opaque black.
 */
internal fun parseCoverHex(hex: String): Int? {
  var s = hex.trim()
  if (s.startsWith("#")) s = s.substring(1)
  val expanded = when (s.length) {
    3 -> buildString(6) { for (c in s) { append(c); append(c) } }
    4 -> buildString(8) { for (c in s) { append(c); append(c) } }
    6, 8 -> s
    else -> return null
  }
  return try {
    val rgba = expanded.toLong(16)
    when (expanded.length) {
      6 -> 0xFF000000.toInt() or rgba.toInt()
      8 -> {
        val r = ((rgba shr 24) and 0xFF).toInt()
        val g = ((rgba shr 16) and 0xFF).toInt()
        val b = ((rgba shr 8) and 0xFF).toInt()
        val a = (rgba and 0xFF).toInt()
        Color.argb(a, r, g, b)
      }
      else -> null
    }
  } catch (_: NumberFormatException) {
    null
  }
}
